up-streamer/src/ustreamer.rs

```diff
@@ -0,0 +1,973 @@
+                    );
+                    error!(
+                        "{}:{}:{} Deleting forwarding rule failed: {:?}",
+                        self.name,
+                        USTREAMER_TAG,
+                        USTREAMER_FN_DELETE_FORWARDING_RULE_TAG,
+                        &no_in_transport_out_authority_found
+                    );
+                    Err(no_in_transport_out_authority_found)
+                }
+            } else {
+                Ok(())
+            }
+        } else {
+            let err = Err(UStatus::fail_with_code(
+                UCode::NOT_FOUND,
+                format!(
+                    "{} is not routed. No rule to delete.",
+                    Self::forwarding_id(&r#in, &out)
+                ),
+            ));
+            error!(
+                "{}:{}:{} Deleting forwarding rule failed: {:?}",
+                self.name, USTREAMER_TAG, USTREAMER_FN_DELETE_FORWARDING_RULE_TAG, err
+            );
+            err
+        }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct ComparableTransport {
+    transport: Arc<Box<dyn UTransport>>,
+}
+
+impl ComparableTransport {
+    pub fn new(transport: Arc<Box<dyn UTransport>>) -> Self {
+        Self { transport }
+    }
+}
+
+impl Hash for ComparableTransport {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Arc::as_ptr(&self.transport).hash(state);
+    }
+}
+
+impl PartialEq for ComparableTransport {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.transport, &other.transport)
+    }
+}
+
+impl Eq for ComparableTransport {}
+
+const TRANSPORT_FORWARDER_TAG: &str = "TransportForwarder:";
+const TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG: &str = "message_forwarding_loop():";
+pub(crate) struct TransportForwarder {}
+
+impl TransportForwarder {
+    async fn new(
+        out_transport: Arc<Box<dyn UTransport>>,
+        message_receiver: Receiver<Arc<UMessage>>,
+    ) -> Self {
+        let out_transport_clone = out_transport.clone();
+        let message_receiver_clone = message_receiver.clone();
+        task::spawn(Self::message_forwarding_loop(
+            UUIDBuilder::build().to_hyphenated_string(),
+            out_transport_clone,
+            message_receiver_clone,
+        ));
+
+        Self {}
+    }
+
+    async fn message_forwarding_loop(
+        id: String,
+        out_transport: Arc<Box<dyn UTransport>>,
+        message_receiver: Receiver<Arc<UMessage>>,
+    ) {
+        while let Ok(msg) = message_receiver.recv().await {
+            debug!(
+                "{}:{}:{} Attempting send of message: {:?}",
+                id,
+                TRANSPORT_FORWARDER_TAG,
+                TRANSPORT_FORWARDER_FN_MESSAGE_FORWARDING_LOOP_TAG,
+                msg
+            );
+
+            let send_res = out_transport.send(msg.deref().clone()).await;
+            if let Err(err) = send_res {
+                error!(
```

Comment:
Fair -- not a threat to streamer itself, but still means we weren't able to forward the message. Perhaps `warn` level?

Reply:
The question is: who will give a damn if this is written to the log (file?)? Is this a problem that needs to be detected during runtime, i.e. raise an alert somewhere? If not, then I guess it will only ever be relevant during development when debugging an application ...

Reply:
Good questions.

I imagined that we'd be logging activity when launching a uStreamer s.t. those logs could be used to find issues with either uEs or the uStreamer itself.

So if a uE which is supposed to be receiving a message _is not_, that's a cause for concern and we could look through the log file to see why the message was not able to be successfully sent over the UTransport.

Reply:
Of course, this is always the purpose of logging. However, I wonder if in a vehicle we have lots of (file system) space available for keeping extensive logs. Maybe that is something that we can discuss in one of the upcoming uProtocol meetings ...

Reply:
Hey @sophokles73, sorry for leaving this hanging, catching back up.

We can and definitely should be mindful of which logging level we're using.

FMPOV I would rather have extensive logging _at the correct levels_, such that when this library is integrated as a part of a uStreamer binary, the appropriate logging level can be set as required.

From past projects, we approached it this way and then would ultimately restrict to say only warn or error levels of logging when deploying for production on vehicle. But to your point, if this were deployed somewhere with enough disk space perhaps they would want up to say info.

WDYT?

Reply:
The situation you should not end up in is having the logs spilled with ERROR/WARNING messages which have been handled without any problems because they indicate _expected_ situations. Just because a situation arises in which something unusual happens, doesn't mean that the situation is worth mentioning. IMHO an ERROR message needs to indicate a situation that we have not expected (and do not know how to properly handle), while a WARNING is something that might endanger the proper functioning of the system if not addressed by somebody. If you agree, you might want to reconsider the usage of WARN/ERROR levels in your code.

Reply:
Sure, I'll take another look through and think about it.

Reply:
I'm going to track this in #15